File: 2025/day03.py
partie1 = 0
partie2 = 0
data = []

with open("2025/input/day03") as f:
    input = f.read().splitlines()

    for line in input:
        data.append([int(x) for x in line])

for row in data:
    x = row.index(max(row[0:-1]))
    a = row[x]
    b = max(row[x + 1 :])

    partie1 += 10*a+b

print("Partie 1:", partie1)

for row in data:
    n = 0
    x = 0
    window = row[:]
    for i in range(12):
        p = 11 - i
        
        if p > 0:
            v = max(window[0:-p])
            n += (10**p) * v
            x = window.index(v) + 1
            window = window[x:]
        else:
            v = max(window)
            n += (10**p) * v
    partie2 += n
        

print("Partie 2:", partie2)

File: 2025/rust/src/aoc2025/day03.rs
use rdcl_aoc_helpers::input::WithReadLines;
use std::fs::File;

fn greedy_number(row: &Vec<u8>, n: u8) -> u64 {
    let mut number: u64 = 0;
    let mut window = row.to_vec();

    for i in 0..n {
        let p = n - i - 1;
        if p > 0 {
            let len = window.len();
            let v = u64::from(*window[..(len - (p as usize))].iter().max().unwrap());
            number += v * (10_u64.pow(p as u32));
            let x = window.iter().position(|&x| u64::from(x) == v).unwrap() + 1;
            window = window[x..].to_vec();
        } else {
            let v: u64 = u64::from(*window.iter().max().unwrap());
            number += v * (10_u64.pow(p as u32));
        }
    }

    return number;
}

pub fn day03(path: &String) {
    let input: Vec<Vec<u8>> = File::open(path)
        .read_lines::<String>(1)
        .into_iter()
        .map(|x| {
            x.chars()
                .map(|x| x.to_string().parse::<u8>().unwrap())
                .collect()
        })
        .collect();

    let mut part1 = 0;
    let mut part2 = 0;

    for row in input {
        part1 += greedy_number(&row, 2);
        part2 += greedy_number(&row, 12);
    }

    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);
}
